### PR TITLE
Optionally export JUnit reports

### PR DESCRIPTION
**Pull request by mcmcgrath13**
* Add JUnit export from `ReTestItems` to `RAITest`
* Beef up `ConcurrentTestSet` with a bit more generic and powerful `RAITestSet`
* Fix precompilation if no `default` RAI config

Test PR: https://github.com/RelationalAI/raicode/pull/13703
Sample datadog report: https://app.datadoghq.com/ci/test-branch/github.com%2FRelationalAI%2Fraicode/raicode/mcm-test-junit?query=test_level%3Atest%20env%3Anone%20%40git.repository.id%3A%22github.com%2FRelationalAI%2Fraicode%22%20%40test.service%3Araicode%20%40git.branch%3Amcm-test-junit&env=none&index=citest&start=1684412198750&end=1685016998750&paused=false

To Do:
- [x] Decide on/implement how to name/place files and whether to just use the `ReTestItems` envs there (probably not?)
- [x] Contribute the `junit_record!` method back to `ReTestItems`
- [x] Confirm this actually works (the guts are here, but haven't gotten it fully running yet)

### CODE REVIEW

File: src/testsets.jl
Comment (mmcgr) [inline src/testsets.jl:13]: Does the junit output need to be part of this, or could it be a separate testset wrapper?
```
@testset JUnitTestSet begin
    @testset RAITestSet begin
...
```
At the cost of more indentation, that would separate responsibility and keep them simpler. If we're generating more than what is contained in a standard testset then this is much less practical.

Comment (mcmcgrath13) [reply src/testsets.jl:13]: The `RAITestSet` is used at all of the non-test_rel levels and is needed for rolling up reports. By default test sets inherit from their parent, so this takes advantage of that, so we only need to mark the outer test set as `RAITestSet` and all the inner ones can be un-annotated.

We could have it just at the root, but then we'd need to maintain another testset type, which is unappealing to me and we'd need to often look to the root of the tree to figure out what's happening and whether to generate junit (and how to record it). The fact that `TestRelTestSet` are nested in `RAITestSet` makes all of this a lot easier.

I'm not sure I'm following the practicality concern - could you expand on that?

Comment (mmcgr) [reply src/testsets.jl:13]: The practicality concerns of keeping it separate that I was thinking of are just what you're describing. RAITestSet doing it all seems the better choice.

**Comment by mcmcgrath13** [comment]
There's still some trailing debug code here (the failures), and a couple minor mysteries left, but I think it's overall ready for review

File: src/testrel.jl
Comment (mcmcgrath13) [inline src/testrel.jl:438]: delete before merge

File: src/testrel.jl
Comment (nickrobinson251) [inline src/testrel.jl:25]: i think we need to do this in `__init__` (i.e. at package _load_ time), otherwise this value will be baked in at _precompile_ time.

e.g. if i don't have a `default` config when i first precompile the package, then i go add a `default` config, then we want `TEXT_CONTEXT` to pick up that new `default` config when the package is next loaded, rather than having to `] precompile` to get the change

(style-wise, i'd normally define `__init__` in the main module file, `src/RAITest.jl`, to reduce the chance of multiple `__init__` functons being defined, but up to you)

```suggestion
function __init__()
    try
        TEST_CONTEXT[] = Context(load_config())
    catch
        @warn "No `default` RAI context found. Use `set_context` to pass in a context and enable usage of `RAITest`."
    end
end
```

File: src/testrel.jl
Comment (nickrobinson251) [inline src/testrel.jl:398]: maybe add a comment on why we need this `name_dict` thing?

File: src/testsets.jl
Comment (nickrobinson251) [inline src/testsets.jl:75]: maybe add a comment why we're doing this here?

presumably the final `finish(ts.dts)` call at the bottom of the function is going to do this (and over-write this value)

(i'm guessing the answer is so that we have the `time_end` set before calling `write_junit_file`, so we can get timings?)

But why not just call `finish(ts.dts)` before the `is_reportable` block? I guess because `finish` also does printing of the results and maybe we want to printing of results _before_ the JUnit writing?

File: src/testsets.jl
Comment (nickrobinson251) [inline src/testsets.jl:151]: similar question here: why is this now needed here?

File: src/testsets.jl
Comment (nickrobinson251) [inline src/testsets.jl:199]: why `TxIDs` not `TxnIDs`? (this isn't my world, do idk what standard abbreviations are, so i'm mostly just this isn't a typo)

**Review by nickrobinson251 — APPROVED** [review]
LGTM (assuming folks who use this package more than me are happy with the `ConcurrentTestSet` -> `RAITestSet` change)

just think we need the `init` change

File: src/testsets.jl
Comment (mcmcgrath13) [reply src/testsets.jl:75]: As you guessed, we need the time end before the writing the junit file. If there are any failures, calling `finish(ts.dts)` throws if it's the root testset, so then we don't get any report. So it's a bit of a dance of getting the data we need to report on and doing reporting before we can call `finish`.

Comment (mcmcgrath13) [reply src/testsets.jl:75]: Also, the way we asynchronously run tests means we lose the connection between a `TestRelTestSet` and its parent, so if we call finish on those, the default test set thinks it's the root and prints the results and throws an error as such. So we don't always finish those test sets.

**Review by mmcgr — APPROVED** [review]
LGTM

**Pull request merged**
